### ISSUE TITLE
[Q]: Which beam-search algorithm to use for real-time transcription using an RNN-T?

### ISSUE THREAD

**Issue by stefan-falk** [espnet/espnet#2408]
Hi! 

I have implemented the RNN-Transducer in Tensorflow and now I am trying to implement a streamable beam-search algorithm. 

For this I've been following `nsc_beam_search`:

https://github.com/espnet/espnet/blob/98d0f5554d3eb11153dc301eb544e7057583ac1c/espnet/nets/beam_search_transducer.py#L413

However, I am not sure if this will allow me to transcribe in real time. The beam-search is already slow even though I am not using a language model (which is also why the quality of the output is rather poor as well at the moment).

My two questions are:

- Is n-step-contrained beam-search a good choice for real-time transcription? 
- What kind of language model can we use in order to improve the results but at the same time keep the execution-time as low as possible?

Thank you if you have any advice on this.

**Comment by b-flo** [comment]
Hi,

> Is n-step-contrained beam-search a good choice for real-time transcription?

It would be if `nstep=1` or even `nstep=2` was used (training and data dependant) and if bashing didn't impact decoding speed. Currently, ALSD, TSD and NSC suffer greatly from the bashing optimization at python level. Current implementation would be faster if prediction net bashing and joint net batching were not used. I'm currently re-thinking implementations for espnet2 to enable bashing with expected performances.

> What kind of language model can we use in order to improve the results but at the same time keep the execution-time as low as possible?

It's so-so from my experiments. LM computation in these algorithms also rely on bashing, thus the extra loops and sorting also impact decoding speed.

Btw, if you use NSC please refer to original OSC and ESPnet (with my tag if possible) for now. NSC is not "publicy" available yet and some modifications are not explained in current code.

**Comment by b-flo** [comment]
I'm closing the issue but feel free to ask further questions if you want!

**Comment by stefan-falk** [comment]
@b-flo Thanks for your response :) 

> It would be if nstep=1 or even nstep=2 was used (training and data dependant) and if batching didn't impact decoding speed ... It would be faster if prediction net batching and joint net batching were not used in current implementation.

What is the problem with batching? Are we talking CPU or GPU? I hope to be able to deploy this model on a mobile device at some point so batching, the way as it is implemented now, is not parallelized. Is this the problem? I'm just trying to figure out what I'd have to do next to speed things up. :D 

Thanks for linking https://github.com/espnet/espnet/pull/2082#issuecomment-678666337 and https://github.com/espnet/espnet/pull/2082#discussion_r479770574.

I have merged the tables here for:

Algorithm [beam=10] | Average time for one utt (s) | CER (test) | WER (test)
-- | -- | -- | --
Default (old) | 1.09 | - | -
Default (new) | 0.47 | 18.6 | 39.0
NSC [nstep=1, prefix-alpha=1] | 0.39 | 24.3 | 50.0
NSC [nstep=3, prefix-alpha=2] | 1.09 | 18.3 | 38.7
ALSD [u-max=20] | 0.36 | 33.2 | 50.8
ALSD [u-max=100] | 0.55 | 18.3 | 38.4
TSD [max_sym_exp=2] | 0.54 | 64.4 | 80.1
TSD [max_sym_exp=4] | 1.18 | 18.2 | 38.5

It looks like none of the implementations show a lot of improvement in turn for a slower execution time, or am I reading this wrong? But that's probably due to the batch-processing you pointed out?

> ALSD, TSD and NSC suffer greatly from the optimization at python level.

It does seem so. It seems that it might be necessary to have a more efficient implementation for this at hand. No way around C++ I guess?

I'd have one additional question regarding NSC:

There is are three nested for-loops which execute the joint-network several times in case a hypothesis is a prefix of another. 

https://github.com/espnet/espnet/blob/98d0f5554d3eb11153dc301eb544e7057583ac1c/espnet/nets/beam_search_transducer.py#L487

I'm not sure if I fully got my head around this part. It seems it re-computes the score for sequences that are prefixes of others but I think I've seen a comment somewhere that stated this part is kind of optional?

**Comment by b-flo** [comment]
> What is the problem with batching? Are we talking CPU or GPU?

It's not related to devices but the way batching is implemented right now. Because we need to handle two types of transitions with/without consuming frames, a lot of extra loops and sorting are needed compared to label-sync decoding. To reach expected decoding speed, we need to handle batching differently (e.g: kick off most loops and time-consuming structures/operations at python level, or directly use C++ and CUDA extensions). Stay tuned, it's in progress!

> I hope to be able to deploy this model on a mobile device at some point so batching, the way as it is implemented now, is not parallelized. Is this the problem? I'm just trying to figure out what I'd have to do next to speed things up. :D

Deployment on mobile devices and RT inference is our goal with transducer models. Outside search algorithms, there are several future additions in work for that.
About parallelization, I suppose you're talking about multiple-utterances batching? If so, it's not a priority right now and it wouldn't be useful for RT/online decoding (or I'm missing some cases).

> It looks like none of the implementations show a lot of improvement in turn for a slower execution time, or am I reading this wrong? But that's probably due to the batch-processing you pointed out?

You're reading the tab right however you didn't take into account the comments below ;)

`It's a bit tricky to compare ALSD to NSC or TSD here. For VIVOS, we need to use the higher values to reach reported performances. However, the lower values for NSC/TSD parameters would be used for other recipes such as Librispeech, while an higher scale for Umax (ALSD) could be needed.`

With a model trained on Librispeech for example, `NSC [nstep=1, prefix-alpha=1]` would be perform better and be faster even with current batching!

> I'm not sure if I fully got my head around this part. It seems it re-computes the score for sequences that are prefixes of others but I think I've seen a comment somewhere that stated this part is kind of optional?

You're right, in the prefix search part we're recomputing scores based on the prefix sequences available in the same set. While it was found optional in default beam search (as I commented somewhere as you said), it was required in NSC to achieve expected performances.

**Comment by stefan-falk** [comment]
@b-flo Hi! 

Sorry for the late reply but I was on vacation last week. :) 

> About parallelization, I suppose you're talking about multiple-utterances batching?

Yes, and I think I understand now, that you were talking about something else. I guess you were talking about parallel execution of the model architecture itself as described in one of the papers, right?

> You're right, in the prefix search part we're recomputing scores based on the prefix sequences available in the same set. While it was found optional in default beam search (as I commented somewhere as you said), it was required in NSC to achieve expected performances.

I see, this probably explains why my results are so bad. It seems that I'll have to implement this part as well then - although it is going to slow down the beam-search a lot.

**Comment by b-flo** [comment]
> Yes, and I think I understand now, that you were talking about something else. I guess you were talking about parallel execution of the model architecture itself as described in one of the papers, right?

Here, batching means we process beam hypotheses in parallel for each utterance while multi-utterances batching means we process multiple utterances at the same time (with beam hypotheses also processed in parallel)

> I see, this probably explains why my results are so bad. It seems that I'll have to implement this part as well then - although it is going to slow down the beam-search a lot.

It will be slowed down but not by a lot. You can also set `prefix-alpha` with a small value to control performance/speed ratio.

**Comment by stefan-falk** [comment]
> Here, batching means we process beam hypotheses in parallel for each utterance while multi-utterances batching means we process multiple utterances at the same time (with beam hypotheses also processed in parallel)

A okay - I misunderstood twice then :D 

> It will be slowed down but not by a lot. You can also set prefix-alpha with a small value to control performance/speed ratio.

I see. I have added the part where I recompute the the score for prefixes but unfortunately the result is absolute garbage and which means I have a bug somewhere. 

May I ask you: Do you spot an obvious error somewhere? To me the code looks similar to the [code from your PR](https://github.com/espnet/espnet/blob/902a534213cd11958323521f40ad98b144dc8dba/espnet/nets/beam_search_transducer.py#L402). The only big difference is that I do not actually compute `n`-steps but just one. I didn't get to this part yet and I assume it should work with `n=1` at least somewhat acceptable ..

May I assume that the implementation from your PR is tested so far in the sense that it produces correct results?


```python
def nsc_beam_search_decode(
    self,
    features: tf.Tensor,
    hypotheses: List[Hypothesis] = None,
    blank=0,
    beam_k=10,
    prefix_alpha=3,
) -> List[Hypothesis]:

    def subtract(x, subset):
        final = []
        for x_ in x:
            if any(x_['yseq'] == sub['yseq'] for sub in subset):
                continue
            final.append(x_)
        return final

    def is_prefix(x, pref):
        if len(pref) >= len(x):
            return False
        for idx in range(len(pref)):
            if pref[idx] != x[idx]:
                return False
        return True

    encoder_outputs, encoder_states = self.encoder_network.call(
        features,
        memory_states=hypotheses[0]['e_memory_states'] if hypotheses else None,
        training=False
    )
    encoder_outputs = encoder_outputs[0]

    if hypotheses is None:
        init_score = tf.constant(0.0, dtype=tf.float32)
        init_yseq = tf.reshape(tf.constant(blank, dtype=tf.int32), (1, 1))
        predict_output, predict_state = self.predict_network.call(init_yseq)
        hypotheses = [
            dict(
                yseq=[blank],
                score=init_score,
                predict_output=[predict_output],
                predict_state=self.predict_network.get_initial_state(),
                e_memory_states=encoder_states
            )
        ]

    for t in range(shape_list(encoder_outputs)[0]):

        A = sorted(hypotheses, key=lambda x: len(x['yseq']), reverse=True)

        encoder_output = tf.reshape(encoder_outputs[t], [1, 1, -1])

        # Re-compute scores for prefixes prefix search

        for j in range(len(A) - 1):
            for i in range((j + 1), len(A)):
                if (
                    is_prefix(A[j]['yseq'], A[i]['yseq'])
                    and (len(A[j]['yseq']) - len(A[i]['yseq'])) <= prefix_alpha
                ):
                    next_id = len(A[i]['yseq'])

                    ytu = tf.nn.log_softmax(self.join_network([encoder_output, A[i]['predict_output'][-1]]), axis=0)
                    ytu = tf.squeeze(ytu)
                    curr_score = float(A[i]['score']) + float(ytu[A[j]['yseq'][next_id]])

                    for k in range(next_id, len(A[j]['yseq']) - 1):
                        ytu = tf.nn.log_softmax(self.join_network.call([encoder_output, A[j]['predict_output'][k]]), axis=0)
                        ytu = tf.squeeze(ytu)
                        curr_score += float(ytu[A[j]['yseq'][k + 1]])

                    A[j]['score'] = np.logaddexp(float(A[j]['score']), curr_score)

        # Calculate all probabilities for possible expansions of all hypotheses

        predict_outputs = tf.concat([hypothesis['predict_output'][-1] for hypothesis in A], axis=0)

        beam_logits = self.join_network.call([encoder_output, predict_outputs])
        beam_log_prob = tf.squeeze(tf.nn.log_softmax(beam_logits, axis=-1), axis=1)
        beam_top_k = tf.nn.top_k(beam_log_prob[..., 1:], k=beam_k)

        Sp = list()
        Vp = list()

        for ti, hypothesis in enumerate(A):

            i_topk = (
                tf.concat([beam_top_k.values[ti], beam_log_prob[ti, ..., 0:1]], -1)[0],
                tf.concat([beam_top_k.indices[ti], [[0]]], axis=-1)[0]
            )

            for logp, k in zip(*i_topk):
                curr_score = float(logp)

                new_hypothesis = dict(
                    yseq=list(hypothesis['yseq']),
                    score=hypothesis['score'] + curr_score,
                    predict_output=hypothesis['predict_output'],
                    predict_state=hypothesis['predict_state'],
                    e_memory_states=hypothesis['e_memory_states']
                )

                if k == blank:
                    Sp.append(new_hypothesis)
                else:
                    new_hypothesis['yseq'].append(int(k + 1))
                    Vp.append(new_hypothesis)

        Vl = sorted(Vp, key=lambda x: x['score'], reverse=True)[:beam_k]
        X = subtract(Vl, A)

        memory_states = _make_batched_memory_states(hypotheses=X, stack_size=len(self.predict_network.lstm_stack))
        pinputs = tf.concat([[[hypothesis['yseq'][-1]]] for hypothesis in X], axis=0)
        pout, pstate = self.predict_network.call(inputs=pinputs, memory_states=memory_states)
        beam_logits = self.join_network.call([encoder_output, pout])

        beam_log_prob = tf.squeeze(tf.nn.log_softmax(beam_logits, axis=-1), axis=1)

        # Compute complete probabilities

        Vp_bar = list()
        for xi, hypothesis in enumerate(X):
            new_hypothesis = dict(
                yseq=list(hypothesis['yseq']),
                score=hypothesis['score'] + beam_log_prob[xi, 0, blank],
                predict_output=hypothesis['predict_output'],
                predict_state=_memory_state_from_batched_states(pstate, xi),
                e_memory_states=hypothesis['e_memory_states']
            )

            new_hypothesis['predict_output'].append(pout[xi:(xi + 1)])
            Vp_bar.append(new_hypothesis)

        hypotheses = sorted(Sp + Vp_bar, key=lambda x: x['score'], reverse=True)[:beam_k]

    return sorted(hypotheses, key=lambda x: x['score'] / len(x['yseq']), reverse=True)
```

**Comment by b-flo** [comment]
> May I ask you: Do you spot an obvious error somewhere? To me the code looks similar to the code from your PR. The only big difference is that I do not actually compute n-steps but just one. I didn't get to this part yet and I assume it should work with n=1 at least somewhat acceptable ..

Sorry, I don't have time to take a closer look right now but will do in a few days! For the one-step version, you can refer to https://arxiv.org/abs/2002.03577, which may help you.

> May I assume that the implementation from your PR is tested so far in the sense that it produces correct results?

All the search algorithms in my PR were tested and validated on several dataset (Librispeech, Voxforge, VIVOS, ESTER and some custom french datasets).

**Comment by stefan-falk** [comment]
@b-flo 

> Sorry, I don't have time to take a closer look right now but will do in a few days! For the one-step version, you can refer to https://arxiv.org/abs/2002.03577, which may help you.

Thank you for the link. I already have this paper in front of me though :) I think it's an issue I am just not seeing. The model "works" using a simple greedy-decode strategy so I am sure that it's just my implementation of NSC beam-search. I totally understand if you have other things to do. Thank you anyway! 👍 

> All the search algorithms in my PR were tested and validated on several dataset (Librispeech, Voxforge, VIVOS, ESTER and some custom french datasets).

Alright, just wanted to make sure that it's definitely my own implementation that's the issue here.

**Comment by b-flo** [comment]
Are outputs totally off or performance is degraded compared to greedy search?
If you have an implementation for default beam search, you could compute number of expansion at each time step. If you have a consequent number of 2+ expansions, I think I know what's the issue (assuming that the algorithm is correct)!

**Comment by stefan-falk** [comment]
@b-flo Well, I just found a bug. I had `predict_output=hypothesis['predict_output']` instead of `predict_output=list(hypothesis['predict_output'])` (or `predict_output=hypothesis['predict_output'][:]`) which did not create a new list and so the wrong list got modified 🤦 

The output is not _completely_ off now anymore. Now I get something like:

```
auf der also warum war die debsituen 
```
instead of:
```
also warum war die psychologie gut
```

But even that is pretty bad compared to greedy decoding which does not produce errors in for the same sample.

It's even worse as I try to stream a longer audio file. Greedy is not perfect but at least gives some reasonable output. My NSC implementation however produces only a few characters and the output does not seem to get longer. ^^

Idk how many times I debugged this code already :D

**Comment by stefan-falk** [comment]
@b-flo Ah, maybe I am the victim of a wrong assumption. In your code you have this part where you check `n < (nstep-1)`.

Since I am not implementing `nstep` yet, just 1-step, I only implemented the else-part:

```python
if n < (nstep - 1):
    for i, v in enumerate(V):
        v.y.append(beam_y[i])

        v.dec_state = decoder.select_state(beam_state, i)

        if rnnlm:
            v.lm_state = select_lm_state(
                beam_lm_states, i, lm_type, lm_layers
            )
            v.lm_scores = beam_lm_scores[i]

    hyps = V[:]
else:
    beam_logp = F.log_softmax(decoder.joint(h_enc, beam_y), dim=-1)

    for i, v in enumerate(V):
        if nstep != 1:
            v.score += float(beam_logp[i, 0])

        v.y.append(beam_y[i])

        v.dec_state = decoder.select_state(beam_state, i)

        if rnnlm:
            v.lm_state = select_lm_state(
                beam_lm_states, i, lm_type, lm_layers
            )
            v.lm_scores = beam_lm_scores[i]
```

That should be what happens for `nstep=1`.

**Comment by b-flo** [comment]
> Ah, maybe I am the victim of a wrong assumption. In your code you have this part where you check n < (nstep-1)

Yes, that's why I said you could use default beam search to compute number of expansions at each time step in your case.

> So, assuming nstep=1 as in my case, this means that I'd not update the score. If it is, I'd update the score according to the probability of the blank symbol (v.score += float(beam_logp[i, 0])).

That's means you'll allow incomplete hypothesis to be carried to the next time step. The condition is not in its complete form for a few reasons, mainly because some parts are missing in ESPnet. But it shouldn't damage performances anyways!

>  Why is that?

I explain all this in a paper that is not yet published, dunno if I should explain all this here ;)

> Also, it looks to me as if we can rewrite this in order to save one decoder.joint call to
> ...

Yes you can!

**Comment by stefan-falk** [comment]
> Yes, that's why I said you could use default beam search to compute number of expansions at each time step in your case.

I re-implemented the `default_beam_search()` from the PR but now I even got trouble to make this run work. There's also this condition I don't quite get.

```python
hyps_max = float(max(hyps, key=lambda x: x.score).score)
kept_most_prob = len(sorted(kept_hyps, key=lambda x: x.score > hyps_max))  # <--

if kept_most_prob >= beam_width:
    break
````

Since we're only sorting, this has no effect and this is equivalent to

```break
if len(kept_hyps) >= beam_width:
  break
```

Is there a chance that this was supposed to be a `filter`?

> I explain all this in a paper that is not yet published, dunno if I should explain all this here ;)

I understand, no worries. :) Is there a release date for the paper? I would be interested in reading it when it's out.

**Comment by b-flo** [comment]
> There's also this condition I don't quite get.
> ...
> Since we're only sorting, this has no effect and this is equivalent to
> ...

Oups. You're right thank you! It should be something like this instead:

```
hyps_max = float(max(hyps, key=lambda x: x.score).score)
kept_most_prob = len(sorted([hyp for hyp in kept_hyps if hyp.score > hyps_max], key=lambda x: x.score))

if kept_most_prob >= beam_width:
    break
```

> I understand, no worries. :) Is there a release date for the paper? I would be interested in reading it when it's out.

In a month I guess! After current PR is merged I'll resume porting transducer implementation to espnet2 and re-write the batching part (or the whole process) to reach expected execution time. Last part is critical for paper release imo.

**Comment by stefan-falk** [comment]
> You're right thank you! It should be.. 

You're welcome :) 

> In a month I guess! After current PR is merged ..

I see. Looking forward to reading it :)

**Comment by stefan-falk** [comment]
@b-flo I'm just not able to make this work .. I don't know how many times I compared this code to your implementation but it seems that I keep overlooking the issue. 😞 

The output is not random but usually just a short string of words. I don't see where my implementation goes wrong. 

This is the updated code I am at right now in case you find the time to take a look. In German we have this expression for it: "_betriebsblind_" (adj.) it's when you become blind for issues or problems on things you're working on or at for too long. 😄 

```python
def nsc_beam_search(
    self,
    features: tf.Tensor,
    hypotheses: List[Hypothesis] = None,
    blank=0,
    beam_width=1,
    beam_k=20,
    prefix_alpha=2,
    nstep=1,
) -> List[Hypothesis]:

    def subtract(x, subset):
        final = []
        for x_ in x:
            if any(x_.yseq == sub.yseq for sub in subset):
                continue
            final.append(x_)
        return final

    def is_prefix(x, pref):
        if len(pref) >= len(x):
            return False
        for idx in range(len(pref)):
            if pref[idx] != x[idx]:
                return False
        return True

    encoder_outputs, encoder_states = self.encoder_network.call(
        features,
        memory_states=hypotheses[0].e_memory_states if hypotheses else None,
        training=False
    )
    encoder_outputs = encoder_outputs[0]

    if hypotheses is None:
        init_score = tf.constant(0.0, dtype=tf.float32)
        init_yseq = tf.reshape(tf.constant(blank, dtype=tf.int32), (1, 1))
        predict_output, predict_state = self.predict_network.call(init_yseq)
        kept_hyps = [
            Hypothesis(
                yseq=[blank],
                score=init_score,
                predict_network_state=PredictNetworkState(
                    predict_output=[predict_output],
                    p_memory_states=self.predict_network.get_initial_state(),
                ),
                e_memory_states=encoder_states
            )
        ]
    else:
        kept_hyps = hypotheses

    for t in range(shape_list(encoder_outputs)[0]):

        encoder_output = tf.reshape(encoder_outputs[t], [1, 1, -1])
        hyps = sorted(kept_hyps, key=lambda x: len(x.yseq), reverse=True)

        # :: Line 5 - 7 | Re-compute scores for prefixes

        for j in range(len(hyps) - 1):
            for i in range((j + 1), len(hyps)):
                if (
                    is_prefix(hyps[j].yseq, hyps[i].yseq)
                    and (len(hyps[j].yseq) - len(hyps[i].yseq)) <= prefix_alpha
                ):
                    next_id = len(hyps[i].yseq)

                    state_i = hyps[i].predict_network_state
                    state_j = hyps[j].predict_network_state

                    ytu = tf.nn.log_softmax(self.join_network([encoder_output, state_i.predict_output[-1]]), axis=0)
                    ytu = tf.squeeze(ytu)
                    curr_score = float(hyps[i].score) + float(ytu[hyps[j].yseq[next_id]])

                    for k in range(next_id, len(hyps[j].yseq) - 1):
                        ytu = tf.nn.log_softmax(
                            self.join_network.call([encoder_output, state_j.predict_output[k]]), axis=0)
                        ytu = tf.squeeze(ytu)
                        curr_score += float(ytu[hyps[j].yseq[k + 1]])

                    hyps[j].score = np.logaddexp(float(hyps[j].score), curr_score)

        S: List[Hypothesis] = list()
        V: List[Hypothesis] = list()

        for n in range(nstep):

            # :: Line 8 - 10 | Expand: Calculate all probabilities for possible expansions of all hypotheses

            predict_outputs = tf.concat(
                [hypothesis.predict_network_state.predict_output[-1] for hypothesis in hyps], axis=0)

            beam_logits = self.join_network.call([encoder_output, predict_outputs])
            beam_log_prob = tf.squeeze(tf.nn.log_softmax(beam_logits, axis=-1), axis=1)
            beam_top_k = tf.nn.top_k(beam_log_prob[..., 1:], k=beam_k)  # Assuming blank has ID 0 (skipped with 1:)

            for ti, hyp in enumerate(hyps):

                ti_top_k = (
                    # Get top-k values concatenating blank
                    tf.concat([beam_top_k.values[ti], beam_log_prob[ti, ..., 0:1]], -1)[0],
                    tf.concat([beam_top_k.indices[ti] + 1, [[0]]], axis=-1)[0]  # +1 because we skipped the blank ID
                )

                for log_prob, k in zip(*ti_top_k):
                    predict_network_state = hyp.predict_network_state
                    new_hypothesis = Hypothesis(
                        yseq=list(hyp.yseq),
                        score=hyp.score + float(log_prob),
                        predict_network_state=PredictNetworkState(
                            predict_output=list(predict_network_state.predict_output),
                            p_memory_states=predict_network_state.p_memory_states,
                        ),
                        e_memory_states=hyp.e_memory_states
                    )

                    if k == blank:
                        # The blank-symbols are not appended
                        S.append(new_hypothesis)
                    else:
                        # Expand with new ID
                        new_hypothesis.yseq.append(int(k))
                        V.append(new_hypothesis)

            # :: Line 11 | Select top hypotheses from Vp (local pruning)

            V = sorted(V, key=lambda x: x.score, reverse=True)

            # :: Line 12 | Compute complete probabilities

            V = subtract(V, hyps)[:beam_width]

            memory_states = _to_batched_states(hypotheses=V, stack_size=len(self.predict_network.lstm_stack))
            pinputs = tf.concat([[[hypothesis.yseq[-1]]] for hypothesis in V], axis=0)
            beam_y, beam_state = self.predict_network.call(inputs=pinputs, memory_states=memory_states)

            if n < (nstep - 1):
                for vi, v in enumerate(V):
                    state = v.predict_network_state
                    state.predict_output.append(beam_y[vi:(vi + 1)])
                    state.p_memory_states = _select_state(beam_state, vi)
                hyps = list(V)
            else:
                beam_logits = self.join_network.call([encoder_output, beam_y])
                beam_log_prob = tf.squeeze(tf.nn.log_softmax(beam_logits, axis=-1), axis=1)
                for vi, v in enumerate(V):
                    state = v.predict_network_state
                    if nstep != 1:
                        v.score += beam_log_prob[vi, 0, blank]
                    state.predict_output.append(beam_y[vi:(vi + 1)])
                    state.p_memory_states = _select_state(beam_state, vi)

        # :: Line 13 | Select top hypotheses from Sp + Vp_bar (global pruning)

        kept_hyps = sorted(S + V, key=lambda x: x.score, reverse=True)[:beam_width]

    return sorted(kept_hyps, key=lambda x: x.score / len(x.yseq), reverse=True)  # [:nbest]

```

**Comment by b-flo** [comment]
Hum, I'm confused by a few things, some may be due to the fact I'm not familiar with tensorflow these days.
It's not really an issue related to ESPnet so I think we should discuss your implementation off the list. it would also be faster to have a textual meeting of some sort and answer our mutual questions. If you can drop me your email, you can delete your message after that.

**Comment by stefan-falk** [comment]
I have my own email-address which I'm able to drop at public places 😆 

It's jon.d@gmx.at 

:)

**Comment by jtkim-kaist** [comment]
Dear @b-flo @stefan-falk,  I'm the author for NSC beam search. NSC is an outdated and pre-mature method (we do not use). Please check our recent research in https://ieeexplore.ieee.org/abstract/document/9250505 and we already implement for on-device ASR based on RNN-T with proposed beam search. Demo: 
https://if.kakao.com/session/69 (at last page)


 Best Regards

**Comment by b-flo** [comment]
Hi @jtkim-kaist,

Thanks a lot for your work! I was aware of adaptive-expansion (AE) search but because of other priorities we didn't consider the addition at that time. i'll definitely check it back later!
Note that what we call `N-step Constrained Beam Search` in ESPnet is our proposition of extension for our original OSC beam search as AE search wasn't implemented/published at that moment ([June 2020](https://github.com/espnet/espnet/pull/2082)) . It's a simple extension of OSC to N-step but we also implemented a special case for N=1 where we allow incomplete hypothesis to be carried onto the next step (not in complete form here). In short, we found out that for some edge cases in time-synchronous decoding (https://ieeexplore.ieee.org/document/9053040) and OSC, enforcing blank transition score after one expansion would result in higely likely hypothesis to be discarded during shuffling/pruning.

I never heard/read of "NSC" outside of ESPnet, do you refer to Adaptive-search algorithm as such? If so, we should may be revisit the naming to avoid confusion.
